docs/decisions/0021-transaction-aggregates.rst

```diff
@@ -0,0 +1,135 @@
+   and less performant, because each of these workflows mush synchronously block on requests from
+   enterprise-access to enterprise-subsidy.
+   
+
+Decision
+========
+Build and maintain *aggregate* redemption models in the enterprise-access service. The aggregated
+metrics we're most interested in right now are total spend (the sum of transaction quantities) and the count
+of transactions.
+The aggregated *dimensions* we're most interested in right now are as follows:
+
+1. count/spend metrics at the **subsidy** level - based on ``subsidy_uuid``
+2. count/spend metrics at the **policy** level - based on ``subsidy_access_policy_uuid``
+3. count/spend metrics at the **(policy, learner)** level - based on (policy id, learner id)
+
+Additionally, we may eventually want aggregates at the **(policy, learner, group)** level, to support
+business record configurations where multiple groups relate to a single policy.
+
+Once these aggregates are in place, replace requests to the enterpirse-subsidy service
+in the ``can_redeem`` and ``redeem`` flows with database queries against these aggregates.
+
+Maintaining these aggregates depends on sourcing transaction and reversal data as follows:
+
+1. Starting from a zero-state in which Aggregate models exist but are not yet populated (backpopulation).
+2. When transactions are committed, any related aggregate records should be updated to reflect it.
+3. When reversals are committed, any related aggregate records should be updated to reflect it.
+
+Inter-service Communication
+---------------------------
+There are several options for how transaction/reversal data can make its way to the aggregates
+layer in the enterprise-access service:
+
+1. Via management commands that reads transaction and reversal records via the enterprise-subsidy
+   REST API.  This approach may work for backpopulation, but is generally disfavorable because
+   it won't help us keep the aggregate records up-to-date in a timely manner (we'd have to run
+   it on some cron/schedule).
+2. Updating the aggregates based on the ``redeem`` response body (on successful redemption).
+   This can help us update aggregates as redemptions occur, but won't help us backpopulate
+   aggregates.  Furthermore, it doesn't help us update aggregates based on any reversal records
+   that are committed.
+3. Emiting transaction/reversal commit events from the enterprise-subsidy service. If we can
+   emit these to an event broker, consumable by enterprise-access, then we can update aggregates
+   as transactions and reversals are committed. Furthermore, we could utilize a Django management
+   command to emit events that already exist (perhaps with ``modified`` times
+   earlier than some adjustable value) as a way to replay history. This would allow us to backpopulate
+   aggregate records.
+
+We'll utilize option (3) to both backpopulate and update aggregate records in real time.
+
+Ensuring Data Consistency
```

Comment:
I'm operating under the assumption that aggregations are calculated initially from a zero state and are thereafter cumulative. 

If this is the case I think we will need some type of job to monitor the aggregation and ensure they are correct. Loops back to my earlier comment that we will need a way to recalculate as well.

Reply:
It's a good idea, added it as another possible mitigation.